h2o-core/src/main/java/water/DKV.java

```diff
@@ -247,23 +247,78 @@ public ClearDKVTask(Key[] retainedKeys) {
     protected void setupLocal() {
       final Set<Key> retainedKeys = new HashSet<>(_retainedKeys.length);
       retainedKeys.addAll(Arrays.asList(_retainedKeys));
+      extractNestedKeys(retainedKeys);
       final Collection<Value> storeKeys = H2O.STORE.values();
       Futures removalFutures = new Futures();
       for (final Value value : storeKeys) {
         if (retainedKeys.contains(value._key)) {
-          if (value.isFrame() || value.isModel()) {
-            continue;
-          } else {
-            Log.info(String.format("Given key %s is of type %d. Not retaining.", value._key.toString(), value.type()));
-          }
+          continue;
         }
         if(value.isNull()) continue;
         if (!value.isModel() && !value.isFrame()) continue;
 
         // It is important to trigger the removal strategy on the Keyed object itself
-        ((Keyed) value.get()).remove(removalFutures);
+        ((Keyed) value.get()).retain(removalFutures, retainedKeys);
       }
       removalFutures.blockForPending();
     }
   }
+
+  /**
+   * Iterates through the keys provided by the user, dropping any keys that are not a Model key or a Frame key.
+   * Afterwards, extracts 
+   *
+   * @param retainedKeys A {@link Set} of retained keys to insert the extracted {@link Frame} and {@link Model} keys to.
+   *                     Should contain user-specified keys to retain in order to extract anything.
+   */
+  private static void extractNestedKeys(final Set<Key> retainedKeys) {
+
+    final Iterator<Key> keysIterator = retainedKeys.iterator(); // Traverse keys provided by the user only.
+    while (keysIterator.hasNext()) {
+      final Key key = keysIterator.next();
+      final Value value = Value.STORE_get(key);
+
+      if (!value.isFrame() && !value.isModel()) {
```

Comment:
could value be null? could the key be invalid?
looking at the code this sure is possible

Reply:
Correct. I'll fix this. Also, the multiple ifs might in fact all be one if-else statement.

Reply:
Fixed in https://github.com/h2oai/h2o-3/pull/3577/commits/391a7ef949c4c82321e1d0cf31f1c227a7c5d096